src/lib/ovalDiscovery.ts

```diff
@@ -0,0 +1,81 @@
+import { StandardChronicleFactory__factory } from "../contract-types/factories/StandardChronicleFactory__factory";
+
+import { StandardPythFactory } from "../contract-types/StandardPythFactory";
+import { StandardPythFactory__factory } from "../contract-types/factories/StandardPythFactory__factory";
+import { FACTORIES_GENESIS_BLOCK } from "./constants";
+import { Logger } from "./";
+
+// Singleton class to discover Oval instances
+export class OvalDiscovery {
+    private static instance: OvalDiscovery;
+    private provider: JsonRpcProvider | undefined;
+    private standardCoinbaseFactory: StandardCoinbaseFactory;
+    private standardChainlinkFactory: StandardChainlinkFactory;
+    private standardChronicleFactory: StandardChronicleFactory;
+    private standardPythFactory: StandardPythFactory;
+    private ovalInstances: Set<string> = new Set();
+
+    private constructor() {
+        this.standardCoinbaseFactory = StandardCoinbaseFactory__factory.connect(env.standardCoinbaseFactory);
+        this.standardChainlinkFactory = StandardChainlinkFactory__factory.connect(env.standardChainlinkFactory);
+        this.standardChronicleFactory = StandardChronicleFactory__factory.connect(env.standardChronicleFactory);
+        this.standardPythFactory = StandardPythFactory__factory.connect(env.standardPythFactory);
+    }
+
+    // Singleton pattern to get an instance of WalletManager
+    public static getInstance(): OvalDiscovery {
+        if (!OvalDiscovery.instance) {
+            OvalDiscovery.instance = new OvalDiscovery();
+        }
+        return OvalDiscovery.instance;
+    }
+
+    public async initialize(provider: JsonRpcProvider) {
+        if (this.provider) return;
+        this.provider = provider;
+        this.findOval(FACTORIES_GENESIS_BLOCK);
+    }
+
+    public async findOval(fromBlock: number) {
+        if (!this.provider) return;
+        const lastBlock = await this.provider.getBlockNumber();
+        const factories = [this.standardCoinbaseFactory, this.standardChainlinkFactory, this.standardChronicleFactory, this.standardPythFactory];
+
+        for (const factory of factories) {
+            const searchConfig: EventSearchConfig = {
+                fromBlock,
+                toBlock: lastBlock,
+                maxBlockLookBack: 20000
+            };
+            const ovalDeployments = await paginatedEventQuery(factory.connect(this.provider), factory.filters.OvalDeployed(undefined, undefined, undefined, undefined, undefined, undefined), searchConfig);
+
+            ovalDeployments.forEach((ovalDeployment: EventLog) => {
+                Logger.debug("OvalDiscovery", `Found Oval deployment: ${ovalDeployment.args[1]}`);
+                this.ovalInstances.add(getAddress(ovalDeployment.args[1]));
+            });
+        }
+
+        setTimeout(() => {
+            this.findOval(lastBlock);
+        }, env.ovalDiscoveryInterval * 1000);
```

Comment:
This class is a service that finds Oval instances deployed through the Factories.

This recursive function acts as a subscription to the events in the contracts. Instead of a subscription, we manually fetch the events every X minutes because event subscriptions sometimes cause issues with WebSockets closing or failing.

I think having the logic of the event retrieval contained in this class and abstracted away from the rest of the code makes it simple to reason about.

The initialization of the Oval discovery service is run in the index file at a higher level: https://github.com/UMAprotocol/oval-node/blob/fee339d47244678816f4cb9882cbf49a3593c517/src/index.ts#L57-L58

Knowing this: Would you extract the logic and handle the event retrieval process differently?

Reply:
i would probably at the very least have a function that runs a singular update, and then have a separate call which runs that update in a loop continuously on the class, just to add a bit of composability and testability